src/GraphBLAS-sharp/MatrixAndVector.fs

```diff
@@ -37,32 +39,56 @@ and [<AbstractClass>] Vector<'a when 'a : struct and 'a : equality>() =
     abstract Length: int
     abstract AsArray: 'a[]
 
-    abstract Item: Mask1D<'a> -> Vector<'a> with get, set
-    abstract Item: int * int -> Scalar<'a> with get, set
-    abstract Item: Mask1D<'a> -> Scalar<'a> with set
+    abstract CreateMask: bool -> Mask1D
 
-    abstract Vxm: Matrix<'a> -> Mask1D<'a> -> Semiring<'a> -> Vector<'a>
-    abstract EWiseAdd: Vector<'a> -> Mask1D<'a> -> Semiring<'a> -> Vector<'a>
-    abstract EWiseMult: Vector<'a> -> Mask1D<'a> -> Semiring<'a> -> Vector<'a>
-    abstract Apply: Mask1D<'a> -> UnaryOp<'a, 'b> -> Vector<'b>
+    abstract Item: Mask1D -> Vector<'a> with get, set
+    abstract Item: int -> Scalar<'a> with get, set
+    abstract Item: Mask1D -> Scalar<'a> with set
+
+    abstract Vxm: Matrix<'a> -> Mask1D -> Semiring<'a> -> Vector<'a>
+    abstract EWiseAdd: Vector<'a> -> Mask1D -> Semiring<'a> -> Vector<'a>
+    abstract EWiseMult: Vector<'a> -> Mask1D -> Semiring<'a> -> Vector<'a>
+    abstract Apply: Mask1D -> UnaryOp<'a, 'b> -> Vector<'b>
     abstract Reduce: Monoid<'a> -> Scalar<'a>
 
-    abstract EWiseAddInplace: Vector<'a> -> Mask1D<'a> -> Semiring<'a> -> unit
-    abstract EWiseMultInplace: Vector<'a> -> Mask1D<'a> -> Semiring<'a> -> unit
-    abstract ApplyInplace: Mask1D<'a> -> UnaryOp<'a, 'b> -> unit
+    abstract EWiseAddInplace: Vector<'a> -> Mask1D -> Semiring<'a> -> unit
+    abstract EWiseMultInplace: Vector<'a> -> Mask1D -> Semiring<'a> -> unit
+    abstract ApplyInplace: Mask1D -> UnaryOp<'a, 'b> -> unit
 
     static member inline (+) (x: Vector<'a>, y: Vector<'a>) = x.EWiseAdd y
     static member inline (*) (x: Vector<'a>, y: Vector<'a>) = x.EWiseMult y
     static member inline (+.*) (x: Vector<'a>, y: Matrix<'a>) = x.Vxm y
     static member inline (.+) (x: Vector<'a>, y: Vector<'a>) = x.EWiseAddInplace y
     static member inline (.*) (x: Vector<'a>, y: Vector<'a>) = x.EWiseMultInplace y
 
-and Mask1D<'a when 'a : struct and 'a : equality> =
-    | Mask1D of Vector<'a>
-    | Complemented1D of Vector<'a>
-    | None
+and Mask1D(isNone: bool, indices: int[], length: int, isRegular: bool) =
+    member this.IsNone = isNone
+    member this.IsRegular = isRegular
+
+    member this.Indices = indices
+    member this.Length = length
+
+    member this.Item
+        with get (idx: int) : bool =
+            this.IsNone || this.Indices |> Array.exists ( ( = ) idx) |> ( = ) this.IsRegular
+
+    static member Create (vector: Vector<'a>) = vector.CreateMask true
+    static member Complemented (vector: Vector<'a>) = vector.CreateMask false
+    static member None = Mask1D(true, Array.empty, 0, true)
+
+and Mask2D(isNone: bool, rows: int[], columns: int[], rowCount: int, columnCount: int, isRegular: bool) =
```

Comment:
Does it make sense to use 2 arrays of `int` instead of 1 array of `(int * int)` here?

Reply:
I guess it is because such representation is more convenient and often used for COO format. But I'm not sure that it is a good solution for .NET interface because in such representation relations between parts of pairs are implicit. We implicitly require that both arrays should be ordered appropriately.